reco_utils/recommender/ripplenet/model.py

```diff
@@ -0,0 +1,241 @@
+        return auc, acc
+    
+    def get_feed_dict(self, n_hop, data, ripple_set, start, end):
+        feed_dict = dict()
+        feed_dict[self.items] = data[start:end, 1]
+        feed_dict[self.labels] = data[start:end, 2]
+        for i in range(n_hop):
+            feed_dict[self.memories_h[i]] = [ripple_set[user][i][0] for user in data[start:end, 0]]
+            feed_dict[self.memories_r[i]] = [ripple_set[user][i][1] for user in data[start:end, 0]]
+            feed_dict[self.memories_t[i]] = [ripple_set[user][i][2] for user in data[start:end, 0]]
+        return feed_dict
+
+    def print_metrics_evaluation(self, sess, n_hop, data, ripple_set, batch_size):
+        start = 0
+        auc_list = []
+        acc_list = []
+        while start < data.shape[0]:
+            auc, acc = self.eval(sess, self.get_feed_dict(n_hop, 
+                                                         data, ripple_set,
+                                                          start, start + batch_size))
+            auc_list.append(auc)
+            acc_list.append(acc)
+            start += batch_size
+        return float(np.mean(auc_list)), float(np.mean(acc_list))
+
+    def fit(self,
+            sess, 
+            n_epoch, batch_size, n_hop,
+            train_data, ripple_set, show_loss):
+        sess.run(tf.global_variables_initializer())
+        for step in range(n_epoch):
+            # training
+            np.random.shuffle(train_data)
+            start = 0
+            while start < train_data.shape[0]:
+                _, loss = self.train(sess, 
+                                     self.get_feed_dict(n_hop, 
+                                     train_data, ripple_set, 
+                                     start, start + batch_size))
+                start += batch_size
+                if show_loss:
+                    print('%.1f%% %.4f' % (start / train_data.shape[0] * 100, loss))
+
+            train_auc, train_acc = self.print_metrics_evaluation(sess, n_hop,
+                                                                 train_data, ripple_set,
+                                                                 batch_size)
+
+            print('epoch %d  train auc: %.4f  acc: %.4f'
+                    % (step, train_auc, train_acc))
+
+    def predict(self, sess, 
+    batch_size, n_hop, data, ripple_set):
+        start = 0
+        labels_list = []
+        scores_list = []
+        while start < data.shape[0]:
+            labels, scores = self.return_scores(sess, 
+                                                self.get_feed_dict(n_hop, 
+                                                data, ripple_set, 
+                                                start, start + batch_size))
+            labels_list.append(labels)
+            scores_list.append(scores)
+            start += batch_size
+        
+        return list(np.concatenate(labels_list)), list(np.concatenate(scores_list))
```

Comment:
this is very inefficient, a better way is to initialize a vector to zeros and assign the results to it. Please see`predict_fn` in https://github.com/ilkarman/DeepLearningFrameworks/blob/master/notebooks/Tensorflow_Inference.ipynb